File: backend/src/services/presenceSignal.ts
import { PEDESTRIAN_CLASSES, VEHICLE_CLASSES } from '../constants/actorClasses';
import { SenseCraftFramePayload } from '../types';

const DEFAULT_ACTIVE_WINDOW_MS = 5000;

function resolveWindowMs() {
  const fromEnv = Number(process.env.PERSON_ACTIVE_WINDOW_MS || DEFAULT_ACTIVE_WINDOW_MS);
  if (!Number.isFinite(fromEnv) || fromEnv <= 0) return DEFAULT_ACTIVE_WINDOW_MS;
  return Math.floor(fromEnv);
}

const activeWindowMs = resolveWindowMs();
let lastPedestrianDetectedAtMs: number | null = null;
let lastVehicleDetectedAtMs: number | null = null;

export type PresenceSignalState = {
  personDetected: boolean;
  vehicleDetected: boolean;
  vehicleOnlyDetected: boolean;
  state: 'GREEN' | 'RED' | 'GRAY';
  lastPersonDetectedAt: string | null;
  personExpiresAt: string | null;
  lastVehicleDetectedAt: string | null;
  vehicleExpiresAt: string | null;
  windowMs: number;
};

function buildState(nowMs = Date.now()): PresenceSignalState {
  const personDetected =
    lastPedestrianDetectedAtMs !== null && nowMs - lastPedestrianDetectedAtMs <= activeWindowMs;
  const vehicleDetected =
    lastVehicleDetectedAtMs !== null && nowMs - lastVehicleDetectedAtMs <= activeWindowMs;
  const vehicleOnlyDetected = vehicleDetected && !personDetected;

  const personExpiresAtMs =
    lastPedestrianDetectedAtMs === null ? null : lastPedestrianDetectedAtMs + activeWindowMs;
  const vehicleExpiresAtMs =
    lastVehicleDetectedAtMs === null ? null : lastVehicleDetectedAtMs + activeWindowMs;

  const state: PresenceSignalState['state'] = personDetected
    ? 'GREEN'
    : vehicleOnlyDetected
      ? 'RED'
      : 'GRAY';

  return {
    personDetected,
    vehicleDetected,
    vehicleOnlyDetected,
    state,
    lastPersonDetectedAt:
      lastPedestrianDetectedAtMs === null ? null : new Date(lastPedestrianDetectedAtMs).toISOString(),
    personExpiresAt: personExpiresAtMs === null ? null : new Date(personExpiresAtMs).toISOString(),
    lastVehicleDetectedAt:
      lastVehicleDetectedAtMs === null ? null : new Date(lastVehicleDetectedAtMs).toISOString(),
    vehicleExpiresAt: vehicleExpiresAtMs === null ? null : new Date(vehicleExpiresAtMs).toISOString(),
    windowMs: activeWindowMs
  };
}

export function updatePresenceSignal(payload: SenseCraftFramePayload): PresenceSignalState {
  const hasPedestrian = payload.detections.some((detection) => PEDESTRIAN_CLASSES.has(detection.class_name));
  const hasVehicle = payload.detections.some((detection) => VEHICLE_CLASSES.has(detection.class_name));
  if (hasPedestrian) {
    lastPedestrianDetectedAtMs = Date.now();
  }
  if (hasVehicle) {
    lastVehicleDetectedAtMs = Date.now();
  }
  return buildState();
}

export function getPresenceSignalState(): PresenceSignalState {
  return buildState();
}

File: backend_py/app/presence_signal.py
from __future__ import annotations

import os
import time
from .actor_classes import PEDESTRIAN_CLASSES, VEHICLE_CLASSES
from .models import FramePayload

DEFAULT_ACTIVE_WINDOW_MS = 5000
active_window_ms = int(os.getenv("PERSON_ACTIVE_WINDOW_MS", str(DEFAULT_ACTIVE_WINDOW_MS)))
if active_window_ms <= 0:
    active_window_ms = DEFAULT_ACTIVE_WINDOW_MS

last_pedestrian_detected_at_ms: float | None = None
last_vehicle_detected_at_ms: float | None = None


def _build_state(now_ms: float | None = None) -> dict:
    now_ms = now_ms or (time.time() * 1000)
    person_detected = (
        last_pedestrian_detected_at_ms is not None
        and (now_ms - last_pedestrian_detected_at_ms) <= active_window_ms
    )
    vehicle_detected = (
        last_vehicle_detected_at_ms is not None
        and (now_ms - last_vehicle_detected_at_ms) <= active_window_ms
    )
    vehicle_only_detected = vehicle_detected and not person_detected

    person_expires_ms = (
        None
        if last_pedestrian_detected_at_ms is None
        else last_pedestrian_detected_at_ms + active_window_ms
    )
    vehicle_expires_ms = (
        None if last_vehicle_detected_at_ms is None else last_vehicle_detected_at_ms + active_window_ms
    )

    state = "GREEN" if person_detected else "RED" if vehicle_only_detected else "GRAY"

    return {
        "personDetected": person_detected,
        "vehicleDetected": vehicle_detected,
        "vehicleOnlyDetected": vehicle_only_detected,
        "state": state,
        "lastPersonDetectedAt": None
        if last_pedestrian_detected_at_ms is None
        else _to_iso(last_pedestrian_detected_at_ms),
        "personExpiresAt": None if person_expires_ms is None else _to_iso(person_expires_ms),
        "lastVehicleDetectedAt": None
        if last_vehicle_detected_at_ms is None
        else _to_iso(last_vehicle_detected_at_ms),
        "vehicleExpiresAt": None if vehicle_expires_ms is None else _to_iso(vehicle_expires_ms),
        "windowMs": active_window_ms,
    }


def _to_iso(ts_ms: float) -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(ts_ms / 1000)) + "Z"


def update_presence_signal(payload: FramePayload) -> dict:
    global last_pedestrian_detected_at_ms, last_vehicle_detected_at_ms

    has_pedestrian = any(d.class_name in PEDESTRIAN_CLASSES for d in payload.detections)
    has_vehicle = any(d.class_name in VEHICLE_CLASSES for d in payload.detections)
    now_ms = time.time() * 1000

    if has_pedestrian:
        last_pedestrian_detected_at_ms = now_ms
    if has_vehicle:
        last_vehicle_detected_at_ms = now_ms

    return _build_state(now_ms)


def get_presence_signal_state() -> dict:
    return _build_state()
